Orange/classification/rules.py

```diff
@@ -0,0 +1,947 @@
+            if temp_rule.validity:
+                rules.append(temp_rule)
+
+        return rules
+
+    def refine_rule(self, X, Y, candidate_rule):
+        covered_X = X[candidate_rule.covered_examples]
+        possible_selectors = []
+
+        for i, attribute in enumerate(candidate_rule.domain.attributes):
+            column = covered_X[:, i]
+            if attribute.is_discrete:
+                for val in [int(x) for x in set(column)]:
+                    s1 = Selector(column=i, op="==", value=val)
+                    s2 = Selector(column=i, op="!=", value=val)
+                    possible_selectors.extend([s1, s2])
+            elif attribute.is_continuous:  # TODO: component based
+                column = np.unique(column)
+                dividers = np.array_split(column, min(10, column.shape[0]))
+                dividers = [np.median(smh) for smh in dividers]
+                for val in dividers:
+                    s1 = Selector(column=i, op="<=", value=val)
+                    s2 = Selector(column=i, op=">=", value=val)
+                    possible_selectors.extend([s1, s2])
+
+        possible_selectors = [smh for smh in possible_selectors if
+                              smh not in candidate_rule.selectors]
+
+        new_rules = []
+        (target_class, selectors, domain, prior_class_dist,
+         quality_evaluator, complexity_evaluator, significance_validator,
+         general_validator) = candidate_rule.seed()
+
+        for curr_selector in possible_selectors:
+            copied_selectors = copy(selectors)
+            copied_selectors.append(curr_selector)
+
+            new_rule = Rule(selectors=copied_selectors,
+                            parent_rule=candidate_rule, domain=domain,
+                            prior_class_dist=prior_class_dist,
+                            quality_evaluator=quality_evaluator,
+                            complexity_evaluator=complexity_evaluator,
+                            significance_validator=significance_validator,
+                            general_validator=general_validator)
+
+            new_rule.filter_and_store(X, Y, target_class, td_optimisation=True)
+            # to ensure that the covered_examples matrices are of the
+            # same size throughout the rule_finder iteration
+
+            if new_rule.validity:
+                new_rules.append(new_rule)
+
+        return new_rules
+
+
+class Selector:
+    """
+    Define a single rule condition.
+    """
+    operators = {
+        # discrete, nominal variables
+        '==': operator.eq,
+        '!=': operator.ne,
+
+        # continuous variables
+        '<=': operator.le,
+        '>=': operator.ge
+    }
+
+    def __init__(self, column, op, value):
+        self.column = column
+        self.op = op
+        self.value = value
+
+    def filter_instance(self, x):
+        return self.operators[self.op](x[self.column], self.value)
```

Comment:
If self is a tuple, `return self.OPERATORS[self[1]](x[self[0]], self[2])` will be somewhat faster.